data-prepper-plugins/rds-source/src/main/java/org/opensearch/dataprepper/plugins/source/rds/export/ExportScheduler.java

```diff
@@ -32,17 +39,183 @@ public class ExportScheduler implements Runnable {
+            LOG.error("The snapshot failed to create, it will be retried");
+            closeExportPartitionWithError(exportPartition);
+            return null;
+        }
+
+        final String snapshotId = snapshotInfo.getSnapshotId();
+        try {
+            checkSnapshotStatus(snapshotId);
+        } catch (Exception e) {
+            LOG.warn("Check snapshot status for {} failed", snapshotId, e);
+            sourceCoordinator.giveUpPartition(exportPartition);
+        }
+
+        LOG.info("Creating an export task for db {} from snapshot {}", exportPartition.getDbIdentifier(), snapshotId);
+        String exportTaskId = exportTaskManager.startExportTask(
+                snapshotInfo.getSnapshotArn(), progressState.getIamRoleArn(), progressState.getBucket(),
+                progressState.getPrefix(), progressState.getKmsKeyId(), progressState.getTables());
+
+        if (exportTaskId != null) {
+            LOG.info("Export task id is {}", exportTaskId);
+            progressState.setExportTaskId(exportTaskId);
+            sourceCoordinator.saveProgressStateForPartition(exportPartition, null);
+        } else {
+            LOG.error("The export task failed to create, it will be retried");
+            closeExportPartitionWithError(exportPartition);
+        }
+
+        return exportTaskId;
+    }
+
+    private void closeExportPartitionWithError(ExportPartition exportPartition) {
+        ExportProgressState exportProgressState = exportPartition.getProgressState().get();
+        // Clear current task id, so that a new export can be submitted.
+        exportProgressState.setExportTaskId(null);
+        sourceCoordinator.closePartition(exportPartition, DEFAULT_CLOSE_DURATION, DEFAULT_MAX_CLOSE_COUNT);
+    }
+
+    private String checkSnapshotStatus(String snapshotId) {
+        LOG.debug("Start checking status of snapshot {}", snapshotId);
+        while (true) {
+            SnapshotInfo snapshotInfo = snapshotManager.checkSnapshotStatus(snapshotId);
+            String status = snapshotInfo.getStatus();
+            // Valid snapshot statuses are: available, copying, creating
+            // The status should never be "copying" here
+            if ("available".equals(status)) {
+                LOG.info("Snapshot {} is available.", snapshotId);
+                return status;
+            }
+
+            LOG.debug("Snapshot {} is still creating. Wait and check later", snapshotId);
+            try {
+                Thread.sleep(DEFAULT_CHECK_STATUS_INTERVAL_MILLS);
+            } catch (InterruptedException e) {
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    private String checkExportStatus(ExportPartition exportPartition) {
+        long lastCheckpointTime = System.currentTimeMillis();
+        String exportTaskId = exportPartition.getProgressState().get().getExportTaskId();
+
+        LOG.debug("Start checking the status of export " + exportTaskId);
+        while (true) {
```

Comment:
Do not have an infinite loop like this. I can cause us to get stuck. Some alternative approaches:

1. Use the main loop of the thread. You can refactor this code to use a field which indicates what should happen in that one loop. Even then, we may want some timeout parameter to retest the conditions that got us into this method in the first place.
2. Have a timeout for checking the export status. After this timeout, you can run the main loop again.

Reply:
I fixed the other infinite loop with a timeout but didn't change this one considering 1) this CheckExportStatus method is executed asynchronously in a separate thread, so it shouldn't block the main loop; 2) export time depends on the size of the data and can sometimes take a long time (e.g. > 1 day). I'm not sure what would be a proper timeout value to set here.

Reply:
I see what you mean here. I think this is fine, but for clarity for developers, you can make this an inner class that implements `Runnable`. Then it will be clear that this is a method that runs a loop.